src/ggml-cuda/ggml-cuda.cu

```diff
@@ -1991,6 +1975,50 @@ static __global__ void k_copy_dst_from_contiguous(char * __restrict__ dst_origin
     }
 }
 
+static inline void prepare_row_mappings(ggml_backend_cuda_context& ctx, int64_t n_as, int64_t n_ids,
+                                       const ggml_tensor * ids, std::vector<int>& moe_counts, std::vector<int>& cum_moe_counts,
+                                       ggml_cuda_pool_alloc<mmid_row_mapping>& dev_row_mapping) {
+
+    GGML_ASSERT(moe_counts.empty() && cum_moe_counts.empty());
+
+    auto stream = ctx.stream();
+
+    std::vector<char> ids_host(ggml_nbytes(ids));
+    const char * ids_dev = (const char *) ids->data;
+    CUDA_CHECK(cudaMemcpyAsync(ids_host.data(), ids_dev, ggml_nbytes(ids), cudaMemcpyDeviceToHost, stream));
```

Comment:
This code results in a race condition. At the end of the function `ids_host` is deallocated and `ids_host.data()` becomes a dangling pointer which can no longer be used safely. Crucially, CUDA device code is by design being run asynchronously from the host code. So without an explicit synchronization there is no guarantee whatsoever that the call to `cudaMemcpyAsync` is executed before the vector is deallocated.